docs: historacle spec (#1562)
umee-network/umee: docs/design_docs/011-historacle-pricing.md

## Consequences

The major negative consequence of this is state bloat. Keeping the prices, medians, and standard deviations of each asset supported by the oracle module leads to a lot of bloat.

The major positive is less technical overhead by utilizing an existing module, and allowing for more safety features for `x/leverage`.

### Backwards Compatibility

This will not introduce a new module, and it is relatively backwards compatible. We should allow for governance to disable this feature by voting to set `Stamp Period`, `Prune Period`, and `Median Period` to 0.

### Positive

- Efficient API for the `x/leverage` module to use for safety.
- We can continue listing low-volume assets for collateral.
- Use of an existing module rather than creating a new one.
- Protects against "borrowing + price dump" attacks which are not prevented by disabling low-volume asset collateral

### Negative

- State bloat.
- Additional processing to do during each epoch.

## Further Discussions

1. Can we lessen state bloat without impacting the `x/leverage` module's efficiency?

## Comments

> Originally a 24-hour median was thought to be effective enough to defend against this type of attack, but our modeling shows that we would need 30 days for adequate protection.

Reviewer: where is this model? It sounds that 30 days is really a long. The price may not reflect the reality. Reality is _now_ modulo possible attack.

# Design Doc 011: Historacle Pricing

## Status

Draft

## Abstract

In order to support small-volume assets in a safe manner, we need to be able to calculate the weighted medians of our prices over the past X amount of time (30 days or so). This will allow the leverage module to make decisions around when to allow liquidations if an asset price is considered abnormal.

## Context

Right now, the volume of assets such as Juno, Umee, and Osmo are so low that it would be relatively easy to use the leverage protocol on UMEE to manipulate the prices of these assets and game the system.

The attack goes:

> 1. Account A deposits & collateralizes a large amount of USDC.
> 2. Account A uses collateral to borrow a lot of the low-volume token (here, FOO).
> 3. Account A sells the tokens to Account B, inflating the price of FOO.
> 4. Account B collateralizes inflated FOO, and borrows USDC.
> 5. Account B exits the market for a profit.

Reviewer 2: We might use the classic scenario
```suggestion
1. Attacker deposits & collateralizes a large amount FOO, a token with low volume on exchanges.
2. Attacker spikes the price of FOO on the exchanges by buying a large amount.
3. Attacker borrows USDC using their FOO as collateral at its current (spiked) oracle price. 
4. The price of FOO returns to normal, and the value of the USDC exceeds the value of the FOO collateral.
5. Account B exits the market for a profit.
```

Reviewer 2: There is also another scenario that might be included:

1. Attacker collateralizes USDC and borrows FOO
2. Attacker sells borrowed FOO on exchanges, dumping the price
3. Attacker borrows an even larger amount of FOO using the same collateral, due to the lower price
4. The price of FOO returns to normal, and the value of the FOO exceeds the value of the USDC collateral.
5. Attacker exits the market for a profit.

Author: I think it'd be good to have both. I'll add the 2nd scenario in the comments section.

> In order to avoid this, and to continue with our goal of allowing users to collateralize low-volume assets, we need to have a safety net.

Reviewer 2: ```suggestion
In order to avoid these attacks, and to continue with our goal of allowing users to collateralize and borrow low-volume assets, we need to have a safety net.
```
`these attacks` only if mentioning both scenarios

Currently, this is defended against by disallowing the use of `Umee` as collateral, however, we would like to re-enable that and list other assets with low volume for collateral.

Historacle pricing will provide an API for the leverage module to tell when prices have changed in an abnormal fashion and avoid such events.

## Specification

> Each `Stamp Period`, the `x/oracle` module will now "stamp" the set of exchange rates in memory, and store it until a `Pruning Period` has passed (30 days). Another epoch, `Median Period`, will determine how often the Median and the `Standard Deviation around the Median` are calculated, which will also be stored in memory.

Author: Applied, we might want to also mention `Prune Period` as there are technically 3 epochs.

> Each `Stamp Period`, the `x/oracle` module will now "stamp" the set of exchange rates in memory, and store it until a `Pruning Period` has passed (30 days). Another epoch, `Median Period`, will determine how often the Median and the `Standard Deviation around the Median` are calculated, which will also be stored in memory.

Author: Clarified in 75baef2c954abc7240b5962aaffe257948649865

These values are stored in state in order to avoid the `x/leverage` module from having to calculate them while making decisions around allowable positions for users to take.

### Proposed API

> The leverage protocol will have access to a `Keeper` function from the `x/oracle` module, which will take a `denom`, and will return whether or not the current price of that asset is within the `Standard Deviation around the Median`.
> 

Reviewer 2: ```suggestion
The leverage protocol will have access to the following `keeper` functions from the `x/oracle` module:
- `historicalMedian(denom) (sdk.Dec, error)` returns the median price of an asset in the last `Pruning Period`
- `withinStandardDeviation(denom) (bool, error)` returns whether or not the current price of an asset is within the `Standard Deviation around the Median`.

```
Is this still the plan?

Author: Do you need `historicalMedian` @toteki ? How else are you going to use it other than using `withinStandardDeviation`?

Reviewer 2: The current plan involves ensuring any `MsgBorrow`, `MsgWithdraw`, or `MsgDecollateralize` results in a position that is healthy (under borrow limit) using both current an historical median prices.

For that I need the specific price - the other option (if I just have a boolean `withinStandardDeviation`) would be to disable all three message types, even for tiny borrow/etc amounts, if one of a user's borrowed or collateral assets  returned `false`.

Author: Gotcha, we'll provide both APIs for ya then 🚀

Author: Please see 04c4082c48d642804c438c18de8d985b6850cfec, changed the function names a bit

## Alternative Approaches

### Feeder implementation

This implementation takes the on-chain process of storing historic prices and calculating the median, and takes it off-chain. Validators would then vote on the prices of these assets in a similar fashion to the `x/oracle` voting process.

There are a couple reasons for not doing this:

1. Technical overhead - the x/oracle module would either have to be heavily augmented to deal with this, or we would have to build a new module to deal with the voting process, which would lead to code replication.
2. Latency - validators take time to come to consensus in this model. Currently prices are 5 blocks behind the current APIs, this would make the median another 5 blocks behind that.

## Consequences

The major negative consequence of this is state bloat. Keeping the prices, medians, and standard deviations of each asset supported by the oracle module leads to a lot of bloat.

The major positive is less technical overhead by utilizing an existing module, and allowing for more safety features for `x/leverage`.

### Backwards Compatibility

> This will not introduce a new module, and it is relatively backwards compatible. We should allow for governance to disable this feature by voting to set `Stamp Period`, `Prune Period`, and `Median Period` to 0.

Author: `Stamp Period` is the amount of blocks we wait to record the price of assets in the state machine (ex. every 30 minutes, we save the set of prices). `Prune Period` is the amount of time after which we delete these prices (ex. every 30 days).

> This will not introduce a new module, and it is relatively backwards compatible. We should allow for governance to disable this feature by voting to set `Stamp Period`, `Prune Period`, and `Median Period` to 0.

Author: clarified this in 75baef2c954abc7240b5962aaffe257948649865

### Positive

- Efficient API for the `x/leverage` module to use for safety.
- We can continue listing low-volume assets for collateral.
- Use of an existing module rather than creating a new one.

### Negative

- State bloat.

> - Additional processing to do during each epoch.

Author: wdym?

> - Additional processing to do during each epoch.

Author: Defined the negatives better here in the computational cost section
